Character Creation

feend78 opened the pull request:
This address Issue #6 

Here is a first draft of a character generation menu. I tried to do my best to have it make sense, but I would like to have a few people test it out and give feedback, especially if more can be done to beef up the help messaging.

This branch includes some significant changes to the configuration, most notably changing the MULTISESSION_MODE parameter to 2, which allows players to create more than one character, and logs them into an OOC area upon connecting where they can create new characters with the `@charcreate` command.

Players are not allowed to exit the character generation menu until it is complete, but in the event of a server restart or disconnection, the `@charcreate` command will resume the menu from the appropriate point. Also, the `@ic` command has been overridden to check whether the character has a `chargen_complete` attribute set on it, and launches the menu if it is not, which disallows puppeting of incomplete characters.

_Note: there is currently a bug in Evennia which will throw an error if you connect with a non-superuser account. I have submitted a pull request to address it, but until that is merged, only the superuser can be used for testing._

feend78 committed:
adjustments to archetypes for handling chargen modifications via `mod` property
added finalize_traits to clean up at the end of chargen

feend78 committed:
Merge branch 'master' into chargen

feend78 committed:
added desc property to archetype objects to display archetype info in chargen
converted color characters in races from { to | style

feend78 committed:
skeleton menu nodes

feend78 committed:
Merge branch 'master' into chargen

feend78 committed:
WIP - chargen complete through trait assignment and partial "equip store"

feend78 committed:
Merge branch 'master' into chargen

Conflicts:
	typeclasses/characters.py
	world/archetypes.py

feend78 committed:
first draft of complete chargen menu
removed the at_object_receive hook from chargen room for testing

feend78 committed:
fixed some bugs in chargen menu tree

feend78 committed:
changed settings to MULTISESSION_MODE=2
implemented the chargen menu as part of the @charcreate command
added the missing -2 to rerflex save on Warrior archetype and updated tests accordingly

feend78 committed:
added tests and corresponding changes
removed all cmdset management within menu nodes
added docstrings

feend78 committed:
disallow quitting the menu before it's complete

feend78 committed:
removed unused imports

ergodicbreak commented on commands/chargen.py:
```diff
@@ -0,0 +1,190 @@
+the Ainneve character creation EvMenu.
+
+This new @charcreate command is added to the session cmdset. We also
+create a cmdset with the "Remove" mergtype to remove the original
```
mergtype -> mergetype

ergodicbreak commented on commands/chargen.py:
```diff
@@ -0,0 +1,190 @@
+        start_location = ObjectDB.objects.get_id(settings.START_LOCATION)
+        default_home = ObjectDB.objects.get_id(settings.DEFAULT_HOME)
+        typeclass = settings.BASE_CHARACTER_TYPECLASS
+        permissions = settings.PERMISSION_PLAYER_DEFAULT
```
given the `@ic` command where potentialy you could puppet any object, how does this interact with the assumption of `BASE_CHARACTER_TYPECLASS`?

feend78 replied:
Good point. I can check the typeclass in  `@ic` and just pass any non-character object through to the default `@ic`.

ergodicbreak commented on commands/chargen.py:
```diff
@@ -0,0 +1,190 @@
+            if new_character.db.chargen_complete:
+                self.msg(("{name} has already completed character "
+                          "generation. Use @ic {name} to puppet.").format(
+                            name=new_character.key
```
Should we stick with a consistent string interpolation method, either `%` or `format`?

feend78 replied:
I've consistently used `format` in the code I've written. I began with the original CmdCharCreate code, which used `%`. I'll go ahead and change them.

ergodicbreak commented on typeclasses/chargen.py:
```diff
@@ -0,0 +1,587 @@
+    text += "\n\nNext, select a race for your character."
+
+    help = fill("Your race gives your character certain bonuses and"
+                "detriments, and makes certain 'focuses' available. "
```
Is there a missing space here between "and" and "detriments"?

feend78 replied:
Yeah, there is. I'm on the fence about using `fill` in general here. In the longer menu nodes, I tended to manually format the line breaks.

feend78 committed:
added a check in @ic to prevent calling character generation on non-character objects
converted remaining string formats to format()
fixed typo in docstring

ergodicbreak commented on typeclasses/chargen.py:
```diff
@@ -0,0 +1,587 @@
+    return (text, help), options
+
+
+def menunode_select_archetype(caller, raw_input):
```
I think we should change the `raw_input` parameter to `raw_string`, which avoids potentially confusing newbies with Python's raw_input function and follows EvMenu's own naming convention.

ergodicbreak commented on typeclasses/chargen.py:
```diff
@@ -0,0 +1,587 @@
+    remaining = archetypes.get_remaining_allocation(char.traits)
+
+    text += "Your character's traits influence combat abilities and skills.\n"
+    text += "Type 'help' to see individual skill definitions.\n\n"
```
I think "skill definitions" here should be "trait definitions"?

ergodicbreak commented on typeclasses/chargen.py:
```diff
@@ -0,0 +1,587 @@
+                    "goto": "menunode_allocate_skills"}
+                   for s in skills.ALL_SKILLS]
+
+        def clear_skills(c, i):
```
`(c, i)`  -> `(caller, raw_input)`?

feend78 replied:
Yes. Neither is used in that callback though.

ergodicbreak replied:
Go ahead and change it to be consistent with the others, I don't think Python has a good method of writing unused params.

feend78 replied:
I looked at all of the callbacks throughout the menu, and nearly all of them accept only one argument, `s`, because throughout this menu, `caller` is the session. 

I like that better for callbacks, as they rarely use the argument anyway, and could easily be modified to never use it.

ergodicbreak commented on typeclasses/chargen.py:
```diff
@@ -0,0 +1,587 @@
+                )
+        help = "Choose carefully. Purchases are final."
+
+        def purchase_item(caller):
```
maybe `purchase_item` could be moved out into the `economy` module as well and brought in with `import`

feend78 replied:
The only reason I can think of that we might not want to do that would be if we want the in-game NPC shops to have actual storerooms for their inventory, similar to how it's done in one of the tutorials, or have them just spawn new items as is done here.

ergodicbreak replied:
perhaps`purchase_item` could have a `storeroom` param that specified the behavior? The potential repeated code doesn't feel right but it'd be fine to punt on it for now.

feend78 committed:
a number of changes per the discussion in PR #56

ergodicbreak commented:
@feend78 on an organizational note do we want the menu nodes in `/typeclasses`? 

Also is there some functionality in EvMenu that timeout's on a yes/no dialogue? Sometimes if I wait on a yes/no, or perhaps go to another tab in the browser and return, the yes/no will expire and the previous node text repeats. It's hard to pin down the exact conditions to reproduce and I'm wondering if this is expected.

feend78 commented:
I think I put them in typeclasses/ because I was roughly following the NPC
Shop tutorial, and that was where the example evmenu module was located. It
makes sense to me to have it in world/, if you think that's better.

I think that timeout thing is a known Evennia or web client issue. I think
I saw @Griatch acknowledge it on the Google group shortly before the
holiday break. I didn't worry about it in connection with our menu,
figuring it'd be solved eventually.
On Jan 3, 2016 7:25 PM, "George Oliver" notifications@github.com wrote:

> @feend78 https://github.com/feend78 on an organizational note do we
> want the menu nodes in /typeclasses?
> 
> Also is there some functionality in EvMenu that timeout's on a yes/no
> dialogue? Sometimes if I wait on a yes/no, or perhaps go to another tab in
> the browser and return, the yes/no will expire and the previous node text
> repeats. It's hard to pin down the exact conditions to reproduce and I'm
> wondering if this is expected.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/evennia/ainneve/pull/56#issuecomment-168562793.

feend78 commented:
I found an issue with the `@ic` command. I've fixed it locally, but I also want to write some tests to cover those cases before pushing it. I'm also going to put an exception to launching the menu for the superuser character as well, so to run character creation on it, the `@charcreate` command will have to be used.

Drevarr commented:
Error in menu node 'menunode_examine_and_buy'. If input contains leading or trailing spaces on the equipment menus.

feend78 committed:
fixed a problem with @ic command not correctly launching the menu
moved chargen menu nodes module to world/ along with its tests
fixed leading/trailing spaces problems

feend78 commented:
Thanks @Drevarr. I just pushed a fix for the leading/trailing spaces issue, as well as the `@ic` issue I mentioned above. Thanks for testing!

feend78 committed:
added tests for @ic and for leading/trailing spaces on menu items

feend78 committed:
modified chargen tests to include padded entries at least once on each node

feend78 committed:
modified the at_look hook on the player typeclass to correct the syntax for @charcreate

feend78 committed:
modified settings to allow regular players 5 characters

feend78 committed:
removed the EvTable display for traits/skills/mana allocation
improved help messaging
added 'tra' alias for traits command

ergodicbreak commented:
hey @feend78 , I was reading IRC logs and it looked like there's still ongoing testing with this, just wanted to check in and confirm that was the case. Would you like to merge or wait?

feend78 commented:
Yeah, it's almost there. I have a few other changes to make as a result of yesterday's testing with my coworkers. I think we're pretty much ready to go otherwise, and once they're in, we can merge. 

Looking over the list of issues/comments from yesterday, there is only one or two little formatting clean-up things to do with the menu text. Most notably, I put a fairly long description of the Skills allocation process in the help, but I'm considering putting it directly in the text of the skills allocation node, as the whole +1/-1 counters thing seemed to trip people up the first time through the menu.

There are also a few buggy things they found with some of the equip/items code that I will address in one or more separate PRs. 

In other news: I now have coworkers pestering me to build a combat system so they can fight each other in Ainneve instead of working. LOL

feend78 committed:
fixed layout problem in traits and skills commands
added wealth command
modifications to layout/ordering in some menu nodes
corresponding changes to tests

feend78 commented:
@ergodicbreak, I think I'm satisfied with where this branch is at this point. As you may have seen in IRC logs, there may still be a few cosmetic changes that we want to make, but I think they can be handled separately later. I'll leave it here for another day for you to review again if you like, and we can merge it tomorrow.

feend78 committed:
converted MV trait to 'gauge' type
added at_turn_start hook and ticker call to replenish MV and mana points

feend78 committed:
minor tweak to language on the equipment 'shopping' node

feend78 committed:
one last language change
